core/server/master/src/main/java/alluxio/master/file/loadmanager/LoadManager.java

```diff
@@ -93,70 +121,130 @@ static class Scheduler {
+      while (true) {
+        if (mCurrentSize.get() < CAPACITY) {
+          return;
+        }
+        CommonUtils.sleepMs(1000);
+      }
+    }
+
+    void runLoad(Load load) {
+      BlockBuffer<BlockBatch> blockBuffer = new BlockBuffer<>(load.getPath());
+      List<WorkerNetAddress> addresses = getWorkerAddresses();
 
-      while (blockIterator.hasNextBatch()) {
-        // Get a worker to handle the load task.
-        ExecutionWorkerInfo worker = getNextAvailableWorker();
-        if (worker == null) { // if no workers available, continue
-          continue;
+      addresses.forEach(address -> {
+        try (CloseableResource<BlockWorkerClient> blockWorker
+                     = mFileSystemContext.acquireBlockWorkerClient(address)) {
+          loadBlockBatch(blockWorker.get(), blockBuffer, load.getBandWidth());
+        } catch (IOException e) {
+          // handle IOException
         }
+      });
+      mCurrentSize.decrementAndGet();
+    }
 
-        List<Long> blockIds = blockIterator.getNextBatchBlocks();
-        BlockBatch blockBatch = new BlockBatch(blockIds, getNextBatchId());
+    private void loadBlockBatch(BlockWorkerClient client,
+                                BlockBuffer<BlockBatch> blockBuffer,
+                                long bandWidth) {
+      if (!checkWorker(client)) {
+        return;
+      }
 
-        CompletableFuture<LoadResponse> completableFuture =
-                CompletableFuture.supplyAsync(() -> worker.execute(blockBatch));
-        completableFuture.thenAccept(load::setLoadResponse);
+      BlockBatch blockBatch = blockBuffer.getNextBatchBlocks();
+      if (blockBatch.getBlockIds().isEmpty()) {
+        return;
       }
+
+      LoadRequest loadRequest = LoadRequest
+              .newBuilder()
+              .addFileBlocks(blockBatch.toProto())
+              .setBandwidth(bandWidth)
+              .build(); // todo: create
+      ListenableFuture<LoadResponse> listenableFuture = client.load(loadRequest);
+      Futures.addCallback(listenableFuture, new FutureCallback<LoadResponse>() {
+        @Override
+        public void onSuccess(LoadResponse r) {
+          if (r.hasStatus()) {
+            TaskStatus s = r.getStatus();
+            if (s != TaskStatus.SUCCESS) {
+              List<BlockStatus> blockStatusList = r.getBlockStatusList();
+              Stream<Long> retryableBlocks =
+                      blockStatusList.stream()
+                              .filter(BlockStatus::getRetryable)
+                              .map(BlockStatus::getBlockId);
+
+              Stream<Long> failedBlocks =
+                      blockStatusList.stream()
+                              .filter(blockStatus -> !blockStatus.getRetryable())
+                              .map(BlockStatus::getBlockId);
+              blockBuffer.addToRetry(retryableBlocks.collect(Collectors.toList()));
+              blockBuffer.addFailedBlocks(failedBlocks.collect(Collectors.toList()));
```

Comment:
Why do you add failed blocks to blockBuffer?

Reply:
Will process the failure and either display/send to user or save for internal debugging